File: packing_interface/src/algorithm_templates/python/root/bl.py
import packing_lib
from typing import List, Tuple


class Packing:
    def solve(self, bin_width: int, rectangles: List[Tuple[int, int, int]]) -> List[Tuple[float, float, int, int]]:
        items = packing_lib.expand_items(rectangles)
        items = packing_lib.sort_by_width(items)
        placements = []

        for item in items:
            width = item["width"]
            height = item["height"]

            if width > bin_width:
                raise ValueError(f"Rectangle width {width} exceeds bin width {bin_width}.")

            position = packing_lib.find_bottom_left_position(
                placements,
                bin_width,
                width,
                height,
            )
            if position is None:
                raise ValueError("Unable to place rectangle with Bottom-Left heuristic.")

            x, y = position
            placements.append((x, y, width, height))

        return packing_lib.output_from_placements(bin_width, placements)

File: packing_interface/src/algorithm_templates/cpp/node/bl.cpp
#include "packing_lib.h"
using namespace packing;

class Repacking {
public:
    std::vector<std::tuple<double, double, int, int>> solve(
        int binHeight,
        int binWidth,
        const std::vector<std::tuple<int, int, int>>& rectangles,
        const std::vector<Obstacle>& obstacles
    ) {
        auto items = sort_by_width(expand_items(rectangles));
        std::vector<Obstacle> blockers = obstacles;
        std::vector<std::tuple<double, double, int, int>> placements;

        for (const auto& item : items) {
            auto position = find_bottom_left_position_with_obstacles(blockers, binWidth, binHeight, item.width, item.height);
            if (!position.has_value()) {
                throw std::runtime_error("Unable to place rectangle with BL node packing");
            }

            auto placement = std::make_tuple(position->first, position->second, item.width, item.height);
            placements.push_back(placement);
            append_placement_as_obstacle(blockers, placement);
        }

        return placements;
    }
};
